sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala

```diff
@@ -2739,6 +2752,26 @@ class Dataset[T] private[sql](
   }
 
   /**
+   * Returns true when the Dataset is cached and materialized.
+   *
+   * @group basic
+   * @since 2.3.0
+   */
+  def isMaterialized(): Boolean = {
+    queryExecution.sparkPlan match {
+      case i: InMemoryTableScanExec =>
+        val blockManager = sparkSession.sparkContext.env.blockManager
+        val rdd = i.relation.cachedColumnBuffers
+        val blockIDs = rdd.partitions.indices.map(index => RDDBlockId(rdd.id, index))
+        blockIDs.foreach { bid =>
+          if (blockManager.get(bid).nonEmpty) blockManager.releaseLock(bid) else return false
```

Comment:
I think this won't work for remote blocks. `BlockManager.get` doesn't acquire any locks for remote blocks. Once you release lock on them, you will get exception.

Btw, for remote blocks, it fetches the blocks back, so it's costly for this purpose.